src/main/java/com/arpnetworking/tsdcore/sinks/HttpSinkActor.java

```diff
@@ -270,42 +327,53 @@ private void dispatchPending() {
+        }
 
         final Request request = requestEntry.getRequest();
         _inflightRequestsCount++;
 
+        fireRequest(request, 0);
+    }
+
+
+    private void fireRequest(final Request request, final int attempt) {
         final CompletableFuture<Response> promise = new CompletableFuture<>();
-        metrics.startTimer(_requestLatencyName);
+        final long requestStartTime = System.currentTimeMillis();
         _client.executeRequest(request, new ResponseAsyncCompletionHandler(promise));
         final CompletionStage<Object> responsePromise = promise
                 .handle((result, err) -> {
-                        metrics.stopTimer(_requestLatencyName);
-                        final Object returnValue;
-                        if (err == null) {
-                            final int responseStatusCode = result.getStatusCode();
-                            final int responseStatusClass = responseStatusCode / 100;
-                            for (final int i : STATUS_CLASSES) {
-                                metrics.incrementCounter(
-                                        String.format("%s/%dxx", _responseStatusName, i),
-                                        responseStatusClass == i ? 1 : 0);
-                            }
-                            if (ACCEPTED_STATUS_CODES.contains(responseStatusCode)) {
-                                 returnValue = new PostSuccess(result);
-                            } else {
-                                 returnValue = new PostRejected(request, result);
-                            }
+                    try (Metrics metrics = _metricsFactory.create()) {
+                        metrics.setTimer(_requestLatencyName, System.currentTimeMillis() - requestStartTime, TimeUnit.MILLISECONDS);
+                    }
+                    if (err == null) {
+                        if (ACCEPTED_STATUS_CODES.contains(result.getStatusCode())) {
+                             return new PostSuccess(attempt, result);
                         } else {
-                            returnValue = new PostFailure(request, err);
+                             return new PostRejected(attempt, request, result);
                         }
-                        metrics.incrementCounter(_requestSuccessName, (returnValue instanceof PostSuccess) ? 1 : 0);
-                        metrics.close();
-                        return returnValue;
+                    } else {
+                        return new PostFailure(attempt, request, err);
+                    }
                 });
         PatternsCS.pipe(responsePromise, context().dispatcher()).to(self());
     }
 
+    private void scheduleRetry(final Request request, final int attempt) {
+        final Duration delay = Duration.ofMillis(
+                ThreadLocalRandom.current().nextInt(
+                        Math.min((int) Math.pow(2, attempt), MAXIMUM_DELAY_MILLIS)
```

Comment:
The other thing missing here is that base backoff. The exponential increase, x1, x2, x4, x16, etc. should be separate from the amount of delay (e.g. 50ms, 100ms, 200ms, 800 ms, etc.).

I like the random component to it, this avoids multiple hosts backing off on the same schedule. I'm not as convinced that it should be a random number between the target backoff and maximum backoff. Perhaps, random between `2^(n-1) * base` and `2^n * base` (both subject to `MAXIMUM_DELAY` of course).

You may want to log debug or trace the amount of backoff being used for the attempt.

Second, you need to be able to configure both the base backoff and the maximum backoff on the builder. There should have reasonable values; maybe 60 seconds maximum and 50 millisecond base. @BrandonArp any opinions on those values?

Reply:
My code is a little unclear, the backoff should probably be (random between `0` and `2^n`) * `base` according to https://en.wikipedia.org/wiki/Exponential_backoff#:~:text=Exponential%20backoff%20is%20an%20algorithm,gradually%20find%20an%20acceptable%20rate.

Reply:
I think you're missing the '* base' part of it

Reply:
Yeah it is. I am adding it and you will see in my next commit